### PR TITLE
Replace the restriction to use Repository inside step with a warning

### PR DESCRIPTION
**Pull request by stefannica**
## Describe changes
The restriction to use Repository inside step is an antiquated measure that has become a nuisance. This PR replaces it with a warning and removes some other unneeded code designed to address this same problem.

## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] If I have added an integration, I have updated the [integrations](https://docs.zenml.io/advanced-guide/integrations) table and the [corresponding website section](https://zenml.io/integrations).
- [x] I have added tests to cover my changes.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

### CODE REVIEW

**Review by schustmi — APPROVED** [review]

File: examples/quickstart/notebooks/quickstart.ipynb [hunk 1]
```diff
@@ -435,7 +435,7 @@
     "@step(enable_cache=False)\n",
     "def prediction_service_loader() -> BaseService:\n",
     "    \"\"\"Load the model service of our train_evaluate_deploy_pipeline.\"\"\"\n",
-    "    repo = Repository(skip_repository_check=True)\n",
```
Comment (fa9r) [inline examples/quickstart/notebooks/quickstart.ipynb:438]: why was this removed? Isn't it still nicer to suppress those warning when running the quickstart?

**Review by fa9r — APPROVED** [review]
Love the change! Only major question I have: are there some methods in Repository() that really should not be used from within a step and might lead to weird behavior?

File: examples/quickstart/notebooks/quickstart.ipynb (hunk 1 above)
Comment (fa9r) [reply examples/quickstart/notebooks/quickstart.ipynb:438]: same for the other change below

Comment (stefannica) [reply examples/quickstart/notebooks/quickstart.ipynb:438]: you had a valid point here. I didn't want to expose users to this hidden variable, so I removed it from here. But the correct way to hide it is to detect whether a step has caching enabled _from within the step_ and only show that warning if that is the case. Which is what I did in the last commit.

**Review by fa9r — APPROVED** [review]
